src/txdb.cpp

```diff
@@ -248,6 +249,52 @@ bool CBlockTreeDB::WriteTxIndex(const std::vector<std::pair<uint256, CDiskTxPos>
     return WriteBatch(batch);
 }
 
+bool CBlockTreeDB::ReadAddrIndex(uint160 addrid, std::vector<CExtDiskTxPos> &list) {
+    boost::scoped_ptr<CDBIterator> pcursor(NewIterator());
+
+    uint256 lookupid;
+    {
+        CHashWriter ss(SER_GETHASH, 0);
+        ss << addrid;
+        lookupid = ss.GetHash();
+    }
+
+    pcursor->Seek(std::make_pair(DB_ADDRINDEX, lookupid));
+
+    while (pcursor->Valid()) {
+        std::pair<std::pair<char, uint256>, CExtDiskTxPos> key;
+        if (pcursor->GetKey(key) && key.first.first == DB_ADDRINDEX && key.first.second == lookupid) {
+            list.push_back(key.second);
+        } else {
+            break;
+        }
+        pcursor->Next();
+    }
+    return true;
+}
+
+bool CBlockTreeDB::WriteAddrIndex(const std::vector<std::pair<uint160, CExtDiskTxPos> > &list) {
+    unsigned char foo[0];
+    CDBBatch batch(*this);
+    for (std::vector<std::pair<uint160, CExtDiskTxPos> >::const_iterator it=list.begin(); it!=list.end(); it++) {
+        CHashWriter ss(SER_GETHASH, 0);
+        ss << it->first;
+        batch.Write(std::make_pair(std::make_pair(DB_ADDRINDEX, ss.GetHash()), it->second), FLATDATA(foo));
+    }
+    return WriteBatch(batch, true);
+}
+
+bool CBlockTreeDB::EraseAddrIndex(const std::vector<std::pair<uint160, CExtDiskTxPos> > &list) {
+    unsigned char foo[0];
+    CDBBatch batch(*this);
+    for (std::vector<std::pair<uint160, CExtDiskTxPos> >::const_iterator it=list.begin(); it!=list.end(); it++) {
+        CHashWriter ss(SER_GETHASH, 0);
+        ss << it->first;
+        batch.Erase(std::make_pair(std::make_pair(DB_ADDRINDEX, ss.GetHash()), it->second));
+    }
+    return WriteBatch(batch, true);
+}
+
 bool CBlockTreeDB::WriteFlag(const std::string &name, bool fValue) {
```

Comment:
These 3 function blocks best moved to the end of the file rather than in the middle